pkg/manager/config/manager.go

```diff
@@ -96,103 +77,82 @@ func (srv *ConfigManager) Init(ctx context.Context, addrs []string, cfg config.A
+	wkey := []byte(path.Join(e.basePath, ns, key))
+	logger := e.logger.With(zap.String("component", string(wkey)))
 	retryInterval := 5 * time.Second
 	e.wg.Run(func() {
-		var prevKV *mvccpb.KeyValue
-
-		ticker := time.NewTicker(e.watchInterval)
-		defer ticker.Stop()
-
-		wch := e.etcdClient.Watch(ctx, wkey)
+		wch := e.kv.NewWatchStream()
+		defer wch.Close()
+		for {
+			_, err := wch.Watch(mvcc.AutoWatchID, wkey, getPrefix(wkey), wch.Rev()-1)
+			if err == nil {
+				break
+			}
+			if k := retryInterval * 2; k < e.watchInterval {
+				retryInterval = k
+			}
+			logger.Warn("failed to watch, will try again later", zap.Duration("sleep", retryInterval))
+			time.Sleep(retryInterval)
+		}
 		for {
 			select {
 			case <-ctx.Done():
 				return
-			case <-ticker.C:
-				resp, err := e.kv.Get(ctx, wkey)
-				if err != nil {
-					logger.Warn("failed to poll", zap.Error(err))
-					break
-				}
-				// len == 0 may mean there is no value set yet, do not warn about that
-				if len(resp.Kvs) > 1 {
-					logger.Warn("failed to poll", zap.Error(ErrNoOrMultiResults))
-					break
-				} else if len(resp.Kvs) == 1 {
-					f(logger, &clientv3.Event{
-						Type:   mvccpb.PUT,
-						Kv:     resp.Kvs[0],
-						PrevKv: prevKV,
-					})
-					prevKV = resp.Kvs[0]
-				}
-			case res := <-wch:
-				if res.Canceled {
-					// don't wait for more than the polling interval
-					if k := retryInterval * 2; k < e.watchInterval {
-						retryInterval = k
-					}
-					logger.Warn("failed to watch, will try again later", zap.Error(res.Err()), zap.Duration("sleep", retryInterval))
-					time.Sleep(retryInterval)
-					wch = e.etcdClient.Watch(ctx, wkey, clientv3.WithCreatedNotify())
-					break
-				}
-
+			case res := <-wch.Chan():
 				for _, evt := range res.Events {
 					f(logger, evt)
-					prevKV = evt.Kv
 				}
-
-				// reset the ticker to prevent another tick immediately
-				ticker.Reset(e.watchInterval)
 			}
 		}
 	})
 }
 
 func (e *ConfigManager) get(ctx context.Context, ns, key string) (*mvccpb.KeyValue, error) {
-	resp, err := e.kv.Get(ctx, path.Join(e.basePath, ns, key))
+	resp, err := e.kv.Range(ctx, []byte(path.Join(e.basePath, ns, key)), nil, mvcc.RangeOptions{Rev: 0})
 	if err != nil {
 		return nil, err
 	}
-	if len(resp.Kvs) != 1 {
+	if len(resp.KVs) != 1 {
 		return nil, ErrNoOrMultiResults
 	}
-	return resp.Kvs[0], nil
+	return &resp.KVs[0], nil
 }
 
-func (e *ConfigManager) list(ctx context.Context, ns string, ops ...clientv3.OpOption) ([]*mvccpb.KeyValue, error) {
-	options := make([]clientv3.OpOption, 1, 1+len(ops))
-	options[0] = clientv3.WithPrefix()
-	options = append(options, ops...)
-	resp, err := e.kv.Get(ctx, path.Join(e.basePath, ns), options...)
-	if err != nil {
-		return nil, err
+func getPrefix(key []byte) []byte {
+	end := make([]byte, len(key))
+	copy(end, key)
+	for i := len(end) - 1; i >= 0; i-- {
+		if end[i] < 0xff {
+			end[i] = end[i] + 1
```

Comment:
What does this mean?
If you get the prefix of a string, why not just use string functions?

Reply:
Copy from etcd server source, in short the next `nextLargestKey`. It is not related to string or whatever.